### PR TITLE
ref(uninstall): purge release history by default

### PR DESCRIPTION
**Pull request by bacongobbler**
`helm delete --purge` is used more often than not, so I suggest we should make it the default convention for Helm 3.

Signed-off-by: Matthew Fisher <matt.fisher@microsoft.com>

### CODE REVIEW

**Review by ian-howell — COMMENTED** [review]
I'm not sure it makes sense to keep this flag if it's going to default to `true`. Wouldn't it be more clear to change the flag? Perhaps `--keep-release`?

Although that seems to run counter to the command's description:
```
This command takes a release name, and then uninstalls the release from Kubernetes.
```

**Comment by bacongobbler** [comment]
It still makes sense to keep the flag as people may not want to purge the release. I'm not sure I understand why we'd want to remove it entierely when we're just changing the default value from false to true.

Not sure why we'd want to rename the flag to something else either. `helm delete --purge=false` seems pretty self-explanatory.

**Review by hickeyma — COMMENTED** [review]
In general a delete command usually removes the item fully. As you mentioned @bacongobbler and from comments in the community, most of the time people want to remove the release fully. Good work for raising this. +1

If this is the case then I think you need a flag instead for retaining the release history, something like `keep-history`, similar to what @ian-howell said. IMO, the `purge` flag would now be counter-intuitive.  

Also the command help would need to be updated:
```console
helm delete --help

This command takes a release name, and then deletes the release from Kubernetes.
It removes all of the resources associated with the last release of the chart.```

File: cmd/helm/uninstall.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func newUninstallCmd(c helm.Interface, out io.Writer) *cobra.Command {
 	f := cmd.Flags()
 	f.BoolVar(&o.dryRun, "dry-run", false, "simulate a uninstall")
 	f.BoolVar(&o.disableHooks, "no-hooks", false, "prevent hooks from running during uninstallation")
-	f.BoolVar(&o.purge, "purge", false, "remove the release from the store and make its name free for later use")
+	f.BoolVar(&o.purge, "purge", true, "remove the release from the store and make its name free for later use")
```
Comment (mattfarina) [inline cmd/helm/uninstall.go]: While I agree with the idea of purging by default but, the flag name and text surrounding it don't quite make sense anymore. For the user experience I like Martins idea of making it a `--keep-history` and having docs that align with that.

**Comment by bacongobbler** [comment]
Okay, I've changed the name to `helm uninstall --keep-release`. I'm not sure how to reword the help text for the "new" feature flag, so I'm looking for any suggestions there.

**Review by hickeyma — COMMENTED** [review]
@bacongobbler Some suggestions on the docs..

`Helm —help:`

"uninstall   given a release name, uninstall the release from Kubernetes" —> "uninstall   uninstall a release"

`helm uninstall --help`

"This command takes a release name, and then uninstalls the release from Kubernetes.
It removes all of the resources associated with the last release of the chart."
to:
"This command takes a release name, and then uninstalls the release.
It removes all of the resources from Kubernetes which are associated with the last release of the chart.
It removes all the Helm history for the chart and the release name."

"--purge         remove the release from the store and make its name free for later use"
to:
"--keep-history         retains the chart Helm history and release name"

**Comment by bacongobbler** [comment]
okay, I amended the docstrings incorporating some of that feedback. Thanks!

File: cmd/helm/uninstall.go [hunk 2]
```diff
@@ -27,8 +27,9 @@ import (
 )
 
 const uninstallDesc = `
-This command takes a release name, and then uninstalls the release from Kubernetes.
-It removes all of the resources associated with the last release of the chart.
+This command takes a release name and uninstalls the release from Kubernetes.
+It removes all of the resources associated with the last release of the chart
+as well as the release history, freeing it up for future use.
```
Comment (hickeyma) [inline cmd/helm/uninstall.go]: Would this read better considering that the Helm history might be stored outside of Kubernetes at some stage?

This command takes a release name and uninstalls the release.
It removes all resources from Kubernetes associated with the last release of the chart. It removes the release history as well.

File: cmd/helm/uninstall.go [hunk 3]
```diff
@@ -64,7 +65,7 @@ func newUninstallCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 	f := cmd.Flags()
 	f.BoolVar(&client.DryRun, "dry-run", false, "simulate a uninstall")
 	f.BoolVar(&client.DisableHooks, "no-hooks", false, "prevent hooks from running during uninstallation")
-	f.BoolVar(&client.Purge, "purge", false, "remove the release from the store and make its name free for later use")
+	f.BoolVar(&client.KeepHistory, "keep-history", false, "remove all associated resources and mark the release as deleted, but retain the release history")
```
Comment (hickeyma) [inline cmd/helm/uninstall.go]: Should the description of the flag be what setting the flag provides and not the whole command. So, something like the following:

"retains the release history including the release name"

**Review by hickeyma — CHANGES_REQUESTED** [review]
@bacongobbler Thanks for the update. Have few small suggestions on the descriptions.

File: cmd/helm/uninstall.go (hunk 3 above)
Comment (bacongobbler) [reply cmd/helm/uninstall.go]: Everything in the description is what the flag provides. No other flag marks the release as deleted; they delete the release ledger entirely. I feel it's important to call that out so users aren't confused what the expected behaviour of this flag is supposed to be.

Comment (hickeyma) [reply cmd/helm/uninstall.go]: I thought the `keep-history` flag retains the Helm history.  I would think the `helm uninstall` command remove all associated resources and mark the release as deleted.

To me, the flag is the differentiator.

**Review by adamreese — APPROVED** [review]

**Comment by bacongobbler** [comment]
@hickeyma any further comments? Would love to get this in for alpha.1

**Comment by hickeyma** [comment]
@bacongobbler I wonder if you could take a look again at https://github.com/helm/helm/pull/5283#pullrequestreview-214941162 ? I know there are small points but I am trying to look at it from the user perspective and what describes the capability best to avoid confusion.

File: cmd/helm/uninstall.go (hunk 3 above)
Comment (mattfarina) [reply cmd/helm/uninstall.go]: If I understand it right, with this change `helm uninstall` will delete the resources and remove all of the CRDs/secrets that contain the historical information. It's kind of like cleaning out the windows registry or app data when the app is uninstalled or removing all associated files from the Mac Library. With `--keep-history` those historical CRDs/secrets are kept around if you want to query them later.

File: cmd/helm/uninstall.go (hunk 2 above)
Comment (mattfarina) [reply cmd/helm/uninstall.go]: > Would this read better considering that the Helm history might be stored outside of Kubernetes at some stage?

When would the Helm history be stored outside of the cluster?

**Review by mattfarina — APPROVED** [review]

File: cmd/helm/uninstall.go (hunk 3 above)
Comment (hickeyma) [reply cmd/helm/uninstall.go]: Yes but the flag imo should only describe what it does and that is keep the helm history.
Its a small point and I am ok with merging either way.

**Pull request merged**